### PR TITLE
allow more remote debug scenarios

### PR DESCRIPTION
**Pull request by patricklee2**
new node 10.1.0

### CODE REVIEW

File: 10.1.0/startup/generateStartupCommand.js [hunk 1]
```diff
@@ -61,18 +61,12 @@ if (!startupCommand && !nodeFile) {
     nodeFile = DEFAULTAPP;
 }
 
-if (!startupCommand && nodeFile && fs.existsSync(nodeFile)) {
-    if (process.env.APPSVC_REMOTE_DEBUGGING == "TRUE") {
```
Comment (rramachand21) [inline 10.1.0/startup/generateStartupCommand.js]: why is this removed?

Comment (patricklee2) [reply 10.1.0/startup/generateStartupCommand.js]: if the startupCmd is process.json, the startupCommand will be `node process.json` which doesnt work
this call has to handle .js files .json files

pm2 should start the node-wrapper which has logic to insert --inspect

I'll double check to make sure it works

File: 10.1.0/startup/generateStartupCommand.js [hunk 2]
```diff
@@ -1,78 +1,180 @@
...
+    if (userStartupCommand) {
+        fs.writeFileSync(CUSTOM_STARTUP_CMD_FLAG, "TRUE"); // set CUSTOM_STARTUP_CMD_FLAG for remote debugging
+    }
+    
+    // contains spaces, so multipart command
+    // eg: npm start
+    if (userStartupCommand.indexOf(" ") != -1) {
+        console.log("Found command: "+ userStartupCommand);
+        return userStartupCommand;
+    }
```
Comment (patricklee2) [inline 10.1.0/startup/generateStartupCommand.js]: if AppCommandLine has a space then its a command and execute as is

File: 10.1.0/startup/generateStartupCommand.js [hunk 3]
```diff
@@ -1,78 +1,180 @@
...
+                fs.writeFileSync(PACKAGE_JSON_FLAG, "TRUE"); // set PACKAGE_JSON_FLAG for remote debugging
+                if (packageJsonPath != "package.json") {
+                    var packageJsonDir = packageJsonPath.substring(0, packageJsonPath.indexOf("package.json")-1);
+                    return "npm --prefix=" + packageJsonDir + " start";
+                } else {
+                    return 'npm start';
+                }
+            }
+        }
+    }
```
Comment (patricklee2) [inline 10.1.0/startup/generateStartupCommand.js]: if package.json is found, run with npm start
node wrapper will add inspect argument

File: 10.1.0/startup/generateStartupCommand.js [hunk 4]
```diff
@@ -1,78 +1,180 @@
...
+        }
     }
-}
 
-var nodeFile = startupCommand;
+    // look for *.js
+    if (userStartupCommand.endsWith(".js")) {
+        console.log("Found js file " + userStartupCommand);
+        return getJsStartupCommand(userStartupCommand);
+    }
```
Comment (patricklee2) [inline 10.1.0/startup/generateStartupCommand.js]: if AppCommandLine is *.js found, run with pm2, if debug run with node

File: 10.1.0/startup/generateStartupCommand.js [hunk 5]
```diff
@@ -1,78 +1,180 @@
...
+
+    // otherwise, use the user input
+    if (userStartupCommand) {
+        console.log(userStartupCommand + " is an executable file");
+        // eg: /home/site/wwwroot/run.sh or ./bin/www
+        return userStartupCommand; 
+    }
```
Comment (patricklee2) [inline 10.1.0/startup/generateStartupCommand.js]: if AppCommandLine exist and did not match anything else run it as an executable

File: 10.1.0/startup/generateStartupCommand.js (hunk 2 above)
Comment (dhanvikapila) [reply 10.1.0/startup/generateStartupCommand.js]: How wold a command like so without spaces `./bin/www` or `./bin/start.sh` be processed?

File: 10.1.0/startup/generateStartupCommand.js [hunk 6]
```diff
@@ -1,78 +1,180 @@
...
+                if (typeof json == 'object' && typeof json.script == 'string') {
+                    var nodeFile = json.script; // run the script directly with node
+                    if (process.env.APPSVC_REMOTE_DEBUGGING_BREAK == "TRUE") {	
+                        return "node --inspect-brk=0.0.0.0:" + process.env.APPSVC_TUNNEL_PORT + " " + nodeFile;
+                    } else {
+                        return "node --inspect=0.0.0.0:" + process.env.APPSVC_TUNNEL_PORT + " " + nodeFile;
+                    }
+                } else {
+                    // did not find script tag
+                    return "pm2 start " + processJsonPath + " --no-daemon";
```
Comment (patricklee2) [inline 10.1.0/startup/generateStartupCommand.js]: if we can not find script tag in process.json, run with pm2

File: 10.1.0/startup/generateStartupCommand.js [hunk 7]
```diff
@@ -1,78 +1,180 @@
...
+        } else {
+            configJsPath = userStartupCommand;
+        }
+        if (fs.existsSync(configJsPath)) {
+            console.log("Found script in " + configJsPath)
+            fs.writeFileSync(PROCESS_JSON_FLAG, "TRUE"); // set PROCESS_JSON_FLAG for remote debugging
+
+            if (process.env.APPSVC_REMOTE_DEBUGGING == "TRUE") {
+                // debugging not yet supported
+                return "pm2 start " + configJsPath + " --no-daemon";
```
Comment (dhanvikapila) [inline 10.1.0/startup/generateStartupCommand.js]: Is there a way for customers to specify the `env` that they want run or the app they want to run using the `--only` variables?

File: 10.1.0/startup/generateStartupCommand.js (hunk 2 above)
Comment (patricklee2) [reply 10.1.0/startup/generateStartupCommand.js]: it would be handle at L141
